File: app/src/main/java/com/moka/framework/util/DateUtil.java
package com.moka.framework.util;


import org.jetbrains.annotations.NotNull;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.Locale;


public class DateUtil {

	private static final String dateFormat_yyyyMMdd = "yyyyMMdd";
	private static final String dateFormat_yyyyMM = "yyyyMM";

	private static final SimpleDateFormat dateFormat_getDiffDayCount = new SimpleDateFormat("yyyyMMdd", Locale.getDefault());
	private static final SimpleDateFormat dateFormat_serverTimeString = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());

	/**
	 * Today & Current Date / Time / Hour / Month / DayOfWeek .. etc
	 */

	public static Date getTodayDate() {
		return Calendar.getInstance().getTime();
	}

	public static int getCurrentWeekOfDay() { // return 1 ~ 7 ( 일요일 월 화 수 목 금 토요일 )
		return Calendar.getInstance().get(Calendar.DAY_OF_WEEK);
	}

	public static int getTomorrowWeekOfDay() { // return 1 ~ 7 ( 일요일 월 화 수 목 금 토요일 )
		Calendar calendar = Calendar.getInstance();
		calendar.add(GregorianCalendar.DAY_OF_YEAR, 1);
		return calendar.get(Calendar.DAY_OF_WEEK);
	}

	public static String getTomorrowWeekOfDayString() { // return 1 ~ 7 ( 일요일 월 화 수 목 금 토요일 )
		Calendar calendar = Calendar.getInstance();
		calendar.add(GregorianCalendar.DAY_OF_YEAR, 1);
		switch ( calendar.get(Calendar.DAY_OF_WEEK) ) {
			case Calendar.SUNDAY:
				return "일요일";
			case Calendar.MONDAY:
				return "월요일";
			case Calendar.TUESDAY:
				return "화요일";
			case Calendar.WEDNESDAY:
				return "수요일";
			case Calendar.THURSDAY:
				return "목요일";
			case Calendar.FRIDAY:
				return "금요일";
			default:
				return "토요일";
		}
	}

	public static int getTodayInt_yyyyMMdd() {
		return formatDateToInt_yyyyMMdd(getTodayDate());
	}

	public static String getTodayString_yyyyMMdd() {
		Date date = new Date();
		SimpleDateFormat dateFormat = new SimpleDateFormat(dateFormat_yyyyMMdd, Locale.getDefault());
		return dateFormat.format(date);
	}

	public static String getTodayString_format(String format) {
		Date date = new Date();
		SimpleDateFormat dateFormat = new SimpleDateFormat(format, Locale.getDefault());
		return dateFormat.format(date);
	}

	public static String getCurrentTimeString() {
		Calendar calendar = Calendar.getInstance();
		int hour = calendar.get(Calendar.HOUR_OF_DAY);
		int minute = calendar.get(Calendar.MINUTE);

		return String.format("%02d:%02d", hour, minute);
	}

	public static int getCurrentHour() { // 24시간제
		return Calendar.getInstance().get(Calendar.HOUR_OF_DAY);
	}

	/**
	 * Get Month // Jan: 0 ~ Dec: 11 ( +1 해서 return )
	 */

	public static int getCurrentMonth_yyyyMM() {
		SimpleDateFormat dateFormat = new SimpleDateFormat(dateFormat_yyyyMM, Locale.getDefault());
		return Integer.parseInt(dateFormat.format(getTodayDate()));
	}

	public static int getMonthFromDate_yyyyMM(Date date) {
		SimpleDateFormat dateFormat = new SimpleDateFormat(dateFormat_yyyyMM, Locale.getDefault());
		return Integer.parseInt(dateFormat.format(date));
	}

	public static Date getMonthDiff(Date date, int diff) {
		Calendar calendar = Calendar.getInstance();
		calendar.setTime(date);
		calendar.add(Calendar.MONTH, diff);
		return calendar.getTime();
	}

	/** **/


	/**
	 * Format
	 */

	public static String formatDateToString(Date date, String format) {
		SimpleDateFormat dateFormat = new SimpleDateFormat(format, Locale.getDefault());
		return dateFormat.format(date);
	}

	public static String formatTimestampToString(long timestamp, String format) {
		return formatDateToString(timestampToDate(timestamp), format);
	}

	public static int formatDateToInt_yyyyMMdd(Date date) {
		SimpleDateFormat dateFormat = new SimpleDateFormat(dateFormat_yyyyMMdd, Locale.getDefault());
		return Integer.parseInt(dateFormat.format(date));
	}

	public static int formatDateToInt_yyyyMM(Date date) {
		SimpleDateFormat dateFormat = new SimpleDateFormat(dateFormat_yyyyMM, Locale.getDefault());
		return Integer.parseInt(dateFormat.format(date));
	}

	public static String formatStringToString(String zTime) {
		SimpleDateFormat dateFormat_to = new SimpleDateFormat("yyyy.MM.dd", Locale.getDefault());
		SimpleDateFormat dateFormat_from = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault());
		try {
			return dateFormat_to.format(dateFormat_from.parse(zTime));
		}
		catch ( ParseException e ) {
			return "";
		}
	}

	public static String formatIntDateToString_yyyyMMdd(int date_yyyyMMdd) {
		SimpleDateFormat dateFormat = new SimpleDateFormat("MM월 dd일", Locale.getDefault());
		return dateFormat.format(parseDate(date_yyyyMMdd));
	}

	public static String formatIntDateToString_format(int date_yyyyMMdd, String format) {
		SimpleDateFormat dateFormat = new SimpleDateFormat(format, Locale.getDefault());
		return dateFormat.format(parseDate(date_yyyyMMdd));
	}

	/** **/

	/**
	 * Parse Date ( should return Date )
	 */

	public static Date parseDate(String date_yyyyMMdd) {
		try {
			SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMdd", Locale.getDefault());
			return dateFormat.parse(date_yyyyMMdd);
		}
		catch ( ParseException parseException ) {
			return new Date();
		}
	}

	public static Date parseDate(int dateTime_yyyyMMdd) {
		return parseDate(String.valueOf(dateTime_yyyyMMdd));
	}

	public static Date parseDate(long dateTime_yyyyMMdd) {
		return parseDate(String.valueOf(dateTime_yyyyMMdd));
	}

	public static Date parseTodayDate(int hour, int minute) {
		Calendar cal = Calendar.getInstance();
		cal.set(Calendar.HOUR_OF_DAY, hour);
		cal.set(Calendar.MINUTE, minute);
		return cal.getTime();
	}

	public static Date parseTomorrowDate(int hour, int minute) {
		Calendar cal = Calendar.getInstance();
		cal.add(GregorianCalendar.DAY_OF_YEAR, 1);
		cal.set(Calendar.HOUR_OF_DAY, hour);
		cal.set(Calendar.MINUTE, minute);
		return cal.getTime();
	}

	/**
	 */

	private static String getDiffTimeStringFromMinutes(int diffMinutes) {
		int diffHours = diffMinutes / 60;

		if ( 48 < diffHours ) {

			if ( 30 < diffMinutes / 60 / 24 )
				return diffMinutes / 60 / 24 / 30 + "달전";
			else
				return diffMinutes / 60 / 24 + "일전";
		}
		else if ( 23 < diffHours && 48 > diffHours )
			return "하루전";
		else if ( 1 <= diffHours )
			return diffHours + "시간 전";
		else if ( 1 < diffMinutes )
			return diffMinutes + "분 전";
		else
			return "방금 전";
	}

	@NotNull
	public static String getDiffTimeString(long createdAt) {
		Date time = timestampToDate(createdAt);
		int diffMinutes = getDiffTimeMinutes(time, new Date());

		return getDiffTimeStringFromMinutes(diffMinutes);
	}

	/** **/

	/**
	 * about TimeStamp
	 */

	public static long getTimestampInSecond() {
		return System.currentTimeMillis() / 1000;
	}

	public static long getTimestampInMillis() {
		return System.currentTimeMillis();
	}

	private static final Calendar calendar_timestampToCalendar = Calendar.getInstance();

	public static Calendar timestampToCalendar(long timestamp) {
		synchronized ( calendar_timestampToCalendar ) {

			calendar_timestampToCalendar.setTimeInMillis(timestamp * 1000);
			return calendar_timestampToCalendar;
		}
	}

	private static final Calendar calendar_getTimestampFromDate = Calendar.getInstance();

	public static long getTimestampFromDate(Date date) {
		synchronized ( calendar_getTimestampFromDate ) {

			calendar_getTimestampFromDate.setTime(date);
			return calendar_getTimestampFromDate.getTimeInMillis();
		}
	}

	public static Date timestampToDate(long timestamp) {
		return timestampToCalendar(timestamp).getTime();
	}

	/** **/

	/**
	 * intDate , diff intDate
	 */

	private static final Calendar calendar_getDateFrom = Calendar.getInstance();

	public static Date getDateFrom(Date date, int diff) {
		synchronized ( calendar_getDateFrom ) {
			calendar_getDateFrom.setTime(date);
			calendar_getDateFrom.add(GregorianCalendar.DAY_OF_YEAR, diff);

			return calendar_getDateFrom.getTime();
		}
	}

	public static Date getDateFrom(int date, int diff) {
		return getDateFrom(parseDate(date), diff);
	}

	public static int getIntDateFrom(int date, int diff) {
		return formatDateToInt_yyyyMMdd(getDateFrom(date, diff));
	}

	private static final Calendar calendar_getDiffDateFromToday = Calendar.getInstance();

	public static Date getDiffDateFromToday(int diff) {
		Date todayDate = getTodayDate();

		synchronized ( calendar_getDiffDateFromToday ) {
			calendar_getDiffDateFromToday.setTime(todayDate);
			calendar_getDiffDateFromToday.add(GregorianCalendar.DAY_OF_YEAR, diff);

			return calendar_getDiffDateFromToday.getTime();
		}
	}

	public static int getIntDiffDateFromToday(int diff) {
		return formatDateToInt_yyyyMMdd(getDiffDateFromToday(diff));
	}

	public static double getDiffMinuteEachTimestamp(long preTimestamp, long timestamp) {
		return Math.floor(( timestamp - preTimestamp ) / 60);
	}

	public static int getDiffDayCount(String fromDate, String toDate) {
		try {
			synchronized ( dateFormat_getDiffDayCount ) {
				return getDiffDayCount(dateFormat_getDiffDayCount.parse(fromDate), dateFormat_getDiffDayCount.parse(toDate));
			}
		}
		catch ( ParseException parseException ) {

			return 0;
		}
	}

	public static int getDiffDayCount(int fromDate, int toDate) {
		return getDiffDayCount(String.valueOf(fromDate), String.valueOf(toDate));
	}

	public static int getDiffDayCount(Date fromDate, Date toDate) {
		int fromDateInt = formatDateToInt_yyyyMMdd(fromDate);
		int toDateInt = formatDateToInt_yyyyMMdd(toDate);

		Date fromDateNew = parseDate(fromDateInt);
		Date toDateNew = parseDate(toDateInt);

		return (int) ( ( toDateNew.getTime() - fromDateNew.getTime() ) / ( 24 * 60 * 60 * 1000 ) );
	}

	public static int getDiffTimeMinutes(Date fromDate, Date toDate) {
		return (int) ( ( toDate.getTime() - fromDate.getTime() ) / ( 60 * 1000 ) );
	}

}

File: app/src/main/java/com/moka/framework/util/DateUtil.kt
package com.moka.framework.util


import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*


object DateUtil {

    private val dateFormat_yyyyMMdd = "yyyyMMdd"
    private val dateFormat_yyyyMM = "yyyyMM"

    private val dateFormat_getDiffDayCount = SimpleDateFormat("yyyyMMdd", Locale.getDefault())

    /**
     * Today & Current Date / Time / Hour / Month / DayOfWeek .. etc
     */

    val todayDate: Date
        get() = Calendar.getInstance().time

    // return 1 ~ 7 ( 일요일 월 화 수 목 금 토요일 )
    val currentWeekOfDay: Int
        get() = Calendar.getInstance().get(Calendar.DAY_OF_WEEK)

    // return 1 ~ 7 ( 일요일 월 화 수 목 금 토요일 )
    val tomorrowWeekOfDay: Int
        get() {
            val calendar = Calendar.getInstance()
            calendar.add(GregorianCalendar.DAY_OF_YEAR, 1)
            return calendar.get(Calendar.DAY_OF_WEEK)
        }

    // return 1 ~ 7 ( 일요일 월 화 수 목 금 토요일 )
    val tomorrowWeekOfDayString: String
        get() {
            val calendar = Calendar.getInstance()
            calendar.add(GregorianCalendar.DAY_OF_YEAR, 1)
            when (calendar.get(Calendar.DAY_OF_WEEK)) {
                Calendar.SUNDAY -> return "일요일"
                Calendar.MONDAY -> return "월요일"
                Calendar.TUESDAY -> return "화요일"
                Calendar.WEDNESDAY -> return "수요일"
                Calendar.THURSDAY -> return "목요일"
                Calendar.FRIDAY -> return "금요일"
                else -> return "토요일"
            }
        }

    val todayInt_yyyyMMdd: Int
        get() = formatDateToInt_yyyyMMdd(todayDate)

    val todayString_yyyyMMdd: String
        get() {
            val date = Date()
            val dateFormat = SimpleDateFormat(dateFormat_yyyyMMdd, Locale.getDefault())
            return dateFormat.format(date)
        }

    fun getTodayString_format(format: String): String {
        val date = Date()
        val dateFormat = SimpleDateFormat(format, Locale.getDefault())
        return dateFormat.format(date)
    }

    val currentTimeString: String
        get() {
            val calendar = Calendar.getInstance()
            val hour = calendar.get(Calendar.HOUR_OF_DAY)
            val minute = calendar.get(Calendar.MINUTE)

            return String.format("%02d:%02d", hour, minute)
        }

    // 24시간제
    val currentHour: Int
        get() = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)

    /**
     * Get Month // Jan: 0 ~ Dec: 11 ( +1 해서 return )
     */

    val currentMonth_yyyyMM: Int
        get() {
            val dateFormat = SimpleDateFormat(dateFormat_yyyyMM, Locale.getDefault())
            return Integer.parseInt(dateFormat.format(todayDate))
        }

    fun getMonthFromDate_yyyyMM(date: Date): Int {
        val dateFormat = SimpleDateFormat(dateFormat_yyyyMM, Locale.getDefault())
        return Integer.parseInt(dateFormat.format(date))
    }

    fun getMonthDiff(date: Date, diff: Int): Date {
        val calendar = Calendar.getInstance()
        calendar.time = date
        calendar.add(Calendar.MONTH, diff)
        return calendar.time
    }

    /**  */


    /**
     * Format
     */

    fun formatDateToString(date: Date, format: String): String {
        val dateFormat = SimpleDateFormat(format, Locale.getDefault())
        return dateFormat.format(date)
    }

    fun formatTimestampToString(timestamp: Long, format: String): String {
        return formatDateToString(timestampToDate(timestamp), format)
    }

    fun formatDateToInt_yyyyMMdd(date: Date): Int {
        val dateFormat = SimpleDateFormat(dateFormat_yyyyMMdd, Locale.getDefault())
        return Integer.parseInt(dateFormat.format(date))
    }

    fun formatDateToInt_yyyyMM(date: Date): Int {
        val dateFormat = SimpleDateFormat(dateFormat_yyyyMM, Locale.getDefault())
        return Integer.parseInt(dateFormat.format(date))
    }

    fun formatStringToString(zTime: String): String {
        val dateFormat_to = SimpleDateFormat("yyyy.MM.dd", Locale.getDefault())
        val dateFormat_from = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault())
        try {
            return dateFormat_to.format(dateFormat_from.parse(zTime))
        } catch (e: ParseException) {
            return ""
        }

    }

    fun formatIntDateToString_yyyyMMdd(date_yyyyMMdd: Int): String {
        val dateFormat = SimpleDateFormat("MM월 dd일", Locale.getDefault())
        return dateFormat.format(parseDate(date_yyyyMMdd))
    }

    fun formatIntDateToString_format(date_yyyyMMdd: Int, format: String): String {
        val dateFormat = SimpleDateFormat(format, Locale.getDefault())
        return dateFormat.format(parseDate(date_yyyyMMdd))
    }

    /**  */

    /**
     * Parse Date ( should return Date )
     */

    fun parseDate(date_yyyyMMdd: String): Date {
        try {
            val dateFormat = SimpleDateFormat("yyyyMMdd", Locale.getDefault())
            return dateFormat.parse(date_yyyyMMdd)
        } catch (parseException: ParseException) {
            return Date()
        }

    }

    fun parseDate(dateTime_yyyyMMdd: Int): Date {
        return parseDate(dateTime_yyyyMMdd.toString())
    }

    fun parseDate(dateTime_yyyyMMdd: Long): Date {
        return parseDate(dateTime_yyyyMMdd.toString())
    }

    fun parseTodayDate(hour: Int, minute: Int): Date {
        val cal = Calendar.getInstance()
        cal.set(Calendar.HOUR_OF_DAY, hour)
        cal.set(Calendar.MINUTE, minute)
        return cal.time
    }

    fun parseTomorrowDate(hour: Int, minute: Int): Date {
        val cal = Calendar.getInstance()
        cal.add(GregorianCalendar.DAY_OF_YEAR, 1)
        cal.set(Calendar.HOUR_OF_DAY, hour)
        cal.set(Calendar.MINUTE, minute)
        return cal.time
    }

    /**
     */

    private fun getDiffTimeStringFromMinutes(diffMinutes: Int): String {
        val diffHours = diffMinutes / 60

        if (48 < diffHours) {

            if (30 < diffMinutes / 60 / 24)
                return "${diffMinutes / 60 / 24 / 30} 달전"
            else
                return "${diffMinutes / 60 / 24} 일전"
        }
        else if (23 < diffHours && 48 > diffHours)
            return "하루전"
        else if (1 <= diffHours)
            return diffHours.toString() + "시간 전"
        else if (1 < diffMinutes)
            return diffMinutes.toString() + "분 전"
        else
            return "방금 전"
    }

    fun getDiffTimeString(createdAt: Long): String {
        val time = timestampToDate(createdAt)
        val diffMinutes = getDiffTimeMinutes(time, Date())

        return getDiffTimeStringFromMinutes(diffMinutes)
    }

    /**  */

    /**
     * about TimeStamp
     */

    val timestampInSecond: Long
        get() = System.currentTimeMillis() / 1000

    val timestampInMillis: Long
        get() = System.currentTimeMillis()

    private val calendar_timestampToCalendar = Calendar.getInstance()

    fun timestampToCalendar(timestamp: Long): Calendar {
        synchronized(calendar_timestampToCalendar) {

            calendar_timestampToCalendar.timeInMillis = timestamp * 1000
            return calendar_timestampToCalendar
        }
    }

    private val calendar_getTimestampFromDate = Calendar.getInstance()

    fun getTimestampFromDate(date: Date): Long {
        synchronized(calendar_getTimestampFromDate) {

            calendar_getTimestampFromDate.time = date
            return calendar_getTimestampFromDate.timeInMillis
        }
    }

    fun timestampToDate(timestamp: Long): Date {
        return timestampToCalendar(timestamp).time
    }

    /**  */

    /**
     * intDate , diff intDate
     */

    private val calendar_getDateFrom = Calendar.getInstance()

    fun getDateFrom(date: Date, diff: Int): Date {
        synchronized(calendar_getDateFrom) {
            calendar_getDateFrom.time = date
            calendar_getDateFrom.add(GregorianCalendar.DAY_OF_YEAR, diff)

            return calendar_getDateFrom.time
        }
    }

    fun getDateFrom(date: Int, diff: Int): Date {
        return getDateFrom(parseDate(date), diff)
    }

    fun getIntDateFrom(date: Int, diff: Int): Int {
        return formatDateToInt_yyyyMMdd(getDateFrom(date, diff))
    }

    private val calendar_getDiffDateFromToday = Calendar.getInstance()

    fun getDiffDateFromToday(diff: Int): Date {
        val todayDate = todayDate

        synchronized(calendar_getDiffDateFromToday) {
            calendar_getDiffDateFromToday.time = todayDate
            calendar_getDiffDateFromToday.add(GregorianCalendar.DAY_OF_YEAR, diff)

            return calendar_getDiffDateFromToday.time
        }
    }

    fun getIntDiffDateFromToday(diff: Int): Int {
        return formatDateToInt_yyyyMMdd(getDiffDateFromToday(diff))
    }

    fun getDiffMinuteEachTimestamp(preTimestamp: Long, timestamp: Long): Double {
        return Math.floor(((timestamp - preTimestamp) / 60).toDouble())
    }

    fun getDiffDayCount(fromDate: String, toDate: String): Int {
        try {
            synchronized(dateFormat_getDiffDayCount) {
                return getDiffDayCount(dateFormat_getDiffDayCount.parse(fromDate), dateFormat_getDiffDayCount.parse(toDate))
            }
        } catch (parseException: ParseException) {

            return 0
        }

    }

    fun getDiffDayCount(fromDate: Int, toDate: Int): Int {
        return getDiffDayCount(fromDate.toString(), toDate.toString())
    }

    fun getDiffDayCount(fromDate: Date, toDate: Date): Int {
        val fromDateInt = formatDateToInt_yyyyMMdd(fromDate)
        val toDateInt = formatDateToInt_yyyyMMdd(toDate)

        val fromDateNew = parseDate(fromDateInt)
        val toDateNew = parseDate(toDateInt)

        return ((toDateNew.time - fromDateNew.time) / (24 * 60 * 60 * 1000)).toInt()
    }

    fun getDiffTimeMinutes(fromDate: Date, toDate: Date): Int {
        return ((toDate.time - fromDate.time) / (60 * 1000)).toInt()
    }

}
